applications/DG-Max/Algorithms/DivDGMaxEigenvalue.cpp

```diff
@@ -49,24 +49,190 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+            err = VecDestroyVecs(numberOfEigenvectors_, &eigenvectors_);
+            CHKERRABORT(PETSC_COMM_WORLD, err);
+            err = VecDuplicateVecs(tempVector_, numberOfConvergedEigenpairs,
+                                   &eigenvectors_);
+            CHKERRABORT(PETSC_COMM_WORLD, err);
+        }
+
+        eigenvalues_.resize(numberOfConvergedEigenpairs);
+        for (PetscInt i = 0; i < numberOfConvergedEigenpairs; ++i) {
+            err = EPSGetEigenpair(solver_, i, &eigenvalues_[i], nullptr,
+                                  eigenvectors_[i], nullptr);
+            CHKERRABORT(PETSC_COMM_WORLD, err);
+        }
+    }
+
+   private:
+    void initKShifts() {
+        DGMaxLogger(VERBOSE, "Initializing boundary shifting");
+        DGMax::FaceMatrixKPhaseShiftBuilder<DIM> builder;
+        builder.setMatrixExtractor([&](const Base::Face* face) {
+            const Base::FaceMatrix& faceMatrix = face->getFaceMatrix(
+                DivDGMaxDiscretization<DIM>::FACE_STIFFNESS_MATRIX_ID);
+            LinearAlgebra::MiddleSizeMatrix block1, block2;
+            block1 = faceMatrix.getElementMatrix(Base::Side::LEFT,
+                                                 Base::Side::RIGHT);
+            block2 = faceMatrix.getElementMatrix(Base::Side::RIGHT,
+                                                 Base::Side::LEFT);
+
+            return std::make_pair(block1, block2);
+        });
+        kphaseshifts_ = builder.build(indexing_);
+    }
+
+    void initSolver() {
+        PetscErrorCode err;
+        err = EPSCreate(PETSC_COMM_WORLD, &solver_);
+        CHKERRABORT(PETSC_COMM_WORLD, err);
+        // Setting operators, but these will be set again after k-shifting
+        // Note the order Mx = (1/omega^2) Sx and NOT Sx = omega^2 M x and doing
+        // a spectral transformation.
+        // It may be possible to remove this relic with the current version of
+        // SLEPc.
+        err = EPSSetOperators(solver_, massMatrix_, stiffnessMatrix_);
+        CHKERRABORT(PETSC_COMM_WORLD, err);
+
+        err = EPSSetWhichEigenpairs(solver_, EPS_LARGEST_MAGNITUDE);
+        CHKERRABORT(PETSC_COMM_WORLD, err);
+        DGMaxLogger(INFO, "Eigenvalue solver configured");
+    }
+
+    Utilities::GlobalIndexing indexing_;
+    Utilities::GlobalPetscMatrix stiffnessMatrix_;
+    Utilities::GlobalPetscMatrix massMatrix_;
+    /// Vector used as intermediate storage and sample
+    Utilities::GlobalPetscVector tempVector_;
+
+    /// exp(ikx) shifts used in the stiffness matrix
+    DGMax::KPhaseShifts<DIM> kphaseshifts_;
+
+    LinearAlgebra::SmallVector<DIM> kpoint_;
+
+    /// Eigenvalue solver
+    EPS solver_;
+
+    /// Number of converged eigenpairs
+    PetscInt numberOfConvergedEigenpairs;
+    /// Number of vectors available in eigenvectors_, this may be larger than
+    /// the number of converged eigenvalues.
+    std::size_t numberOfEigenvectors_;
+    /// Storage for eigenvectors
+    Vec* eigenvectors_;
```

Comment:
Unfortunately not that I know. I'm working with [VecDuplicateVecs](https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Vec/VecDuplicateVecs.html), to duplicate a single vector to get an array of vectors of the same size/type/shape. These have to be cleared later by `VecDestroyVecs`.

So the only (and a bit frustrating way) around it would be to store the vectors in say a `std::vector<Vec>` and then duplicate them manually. But that would not remove the need for cleaning them up manually (and then individually).

Reply:
okay

Reply:
https://www.dealii.org/current/doxygen/deal.II/classPETScWrappers_1_1MPI_1_1Vector.html

Reply:
For `Vec` you might be able to, but as you already see here there you would need to distinguish between 1 and N vectors.  But with `Mat` I have seen several places where you are not even allowed to destroy the matrix (because Petsc does that for you). So it completely depends on the matrix/vector/solver type, and how you obtained it, what the correct way of destroying it is. In that light I totally understand this warning given [on the petsc mailing list](https://lists.mcs.anl.gov/pipermail/petsc-users/2015-August/026561.html)

The use case for dealii I totally understand it, because they provide it as alternative to their own vector and matrix classes, so that you can easily swap them out or even swap them for trilinos. Though note that even there it required switching around several things (at least that was my experience). Given their far more manpower and advantage of swapping libraries, I can totally see that a (leaky?) abstraction may be worth it.

So in my view, replacing this pointer by some wrapper class, would only replace one problem (memory management/overruns) by other problems (very leaky wrapper, maintenance cost of the wrapper, mental overhead to deal with the wrapper).

Reply:
Fair point